// angular-tree-component/dist/mobx-angular/tree-mobx-autorun.directive.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Directive, ViewContainerRef, TemplateRef, Input } from '@angular/core';
import { autorun } from 'mobx';
var TreeMobxAutorunDirective = /** @class */ (function () {
    function TreeMobxAutorunDirective(templateRef, viewContainer) {
        this.templateRef = templateRef;
        this.viewContainer = viewContainer;
        this.templateBindings = {};
    }
    TreeMobxAutorunDirective.prototype.ngOnInit = function () {
        this.view = this.viewContainer.createEmbeddedView(this.templateRef);
        if (this.dispose) {
            this.dispose();
        }
        if (this.shouldDetach()) {
            this.view.detach();
        }
        this.autoDetect(this.view);
    };
    TreeMobxAutorunDirective.prototype.shouldDetach = function () {
        return this.treeMobxAutorun && this.treeMobxAutorun.detach;
    };
    TreeMobxAutorunDirective.prototype.autoDetect = function (view) {
        this.dispose = autorun(function () { return view.detectChanges(); });
    };
    TreeMobxAutorunDirective.prototype.ngOnDestroy = function () {
        if (this.dispose) {
            this.dispose();
        }
    };
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], TreeMobxAutorunDirective.prototype, "treeMobxAutorun", void 0);
    TreeMobxAutorunDirective = __decorate([
        Directive({ selector: '[treeMobxAutorun]' }),
        __metadata("design:paramtypes", [TemplateRef,
            ViewContainerRef])
    ], TreeMobxAutorunDirective);
    return TreeMobxAutorunDirective;
}());
export { TreeMobxAutorunDirective };
