// angular-tree-component/dist/angular-tree-component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { TreeMobxAutorunDirective } from './mobx-angular/tree-mobx-autorun.directive';
import { TREE_ACTIONS } from './models/tree-options.model';
import { KEYS } from './constants/keys';
import { TreeModel } from './models/tree.model';
import { TreeNode } from './models/tree-node.model';
import { TreeDraggedElement } from './models/tree-dragged-element.model';
import { TreeVirtualScroll } from './models/tree-virtual-scroll.model';
import { LoadingComponent } from './components/loading.component';
import { TreeComponent } from './components/tree.component';
import { TreeNodeComponent } from './components/tree-node.component';
import { TreeNodeContent } from './components/tree-node-content.component';
import { TreeNodeDropSlot } from './components/tree-node-drop-slot.component';
import { TreeNodeExpanderComponent } from './components/tree-node-expander.component';
import { TreeNodeChildrenComponent } from './components/tree-node-children.component';
import { TreeNodeCollectionComponent } from './components/tree-node-collection.component';
import { TreeNodeWrapperComponent } from './components/tree-node-wrapper.component';
import { TreeViewportComponent } from './components/tree-viewport.component';
import { TreeNodeCheckboxComponent } from './components/tree-node-checkbox.component';
import { TreeDropDirective } from './directives/tree-drop.directive';
import { TreeDragDirective } from './directives/tree-drag.directive';
import { TreeAnimateOpenDirective } from './directives/tree-animate-open.directive';
import './polyfills';
var TreeModule = /** @class */ (function () {
    function TreeModule() {
    }
    TreeModule_1 = TreeModule;
    TreeModule.forRoot = function () {
        return {
            ngModule: TreeModule_1,
            providers: [TreeDraggedElement]
        };
    };
    var TreeModule_1;
    TreeModule = TreeModule_1 = __decorate([
        NgModule({
            declarations: [
                TreeComponent,
                TreeNodeComponent,
                TreeNodeContent,
                LoadingComponent,
                TreeDropDirective,
                TreeDragDirective,
                TreeNodeExpanderComponent,
                TreeNodeChildrenComponent,
                TreeNodeDropSlot,
                TreeNodeCollectionComponent,
                TreeViewportComponent,
                TreeNodeWrapperComponent,
                TreeNodeCheckboxComponent,
                TreeAnimateOpenDirective,
                TreeMobxAutorunDirective
            ],
            exports: [
                TreeComponent,
                TreeNodeComponent,
                TreeNodeContent,
                LoadingComponent,
                TreeDropDirective,
                TreeDragDirective,
                TreeNodeExpanderComponent,
                TreeNodeChildrenComponent,
                TreeNodeDropSlot,
                TreeNodeCollectionComponent,
                TreeViewportComponent,
                TreeNodeWrapperComponent,
                TreeNodeCheckboxComponent,
                TreeAnimateOpenDirective,
                TreeMobxAutorunDirective
            ],
            imports: [CommonModule],
            providers: []
        })
    ], TreeModule);
    return TreeModule;
}());
export { TreeModule };
export { TreeModel, TreeNode, TreeDraggedElement, TreeVirtualScroll, TREE_ACTIONS, KEYS, LoadingComponent, TreeAnimateOpenDirective, TreeComponent, TreeNodeComponent, TreeNodeWrapperComponent, TreeNodeContent, TreeDropDirective, TreeDragDirective, TreeNodeExpanderComponent, TreeNodeChildrenComponent, TreeNodeDropSlot, TreeNodeCollectionComponent, TreeViewportComponent, TreeNodeCheckboxComponent };
